### PR TITLE
Geom: Implement `Line` class

### PR DESCRIPTION
**Pull request by jsn0716**
### Description
Added the `Line` class to `new_geom`

### Testing Done
Added unit tests in `line_test.cpp`

### Resolved Issues
Part 3/5 for #957 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/new_geom/line.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     * @param coeff the coefficient
+     *
+     * @throws std::invalid_argument if order > 1
+     */
+    inline void setCoeff(unsigned int order, double coeff) override
+    {
+        if (order > 1)
+        {
+            throw std::invalid_argument(
+                "Tried to set the coefficient of term with order greater than 1");
```
Comment (scveloso) [inline src/software/new_geom/line.h]: Might be good to indicate that we tried to do this for a ```Line``` and that its not allowed

**Review by scveloso — COMMENTED** [review]
LGTM just one minor comment

File: src/software/new_geom/line.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    inline explicit Line() {}
+
+    /**
+     * Creates a line from a y-intercept and a slope
+     *
+     * @param y_intercept the y_intercept of the line
+     * @param slope the slope of the line
+     */
+    inline explicit Line(double y_intercept, double slope)
```
Comment (MathewMacDougall) [inline src/software/new_geom/line.h]: What was the reason for inlining the class? I'd generally prefer a separate `.cpp` for consistency

File: src/software/new_geom/line.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "software/new_geom/polynomial.h"
+
+class Line final : public Polynomial
+{
+   public:
+    /**
+     * Creates a line with zero y-intercept and zero slope
+     */
+    inline explicit Line() {}
```
Comment (MathewMacDougall) [inline src/software/new_geom/line.h]: Can we also have a constructor that creates a line given 2 `Points`?

File: src/software/new_geom/line.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <stdexcept>
+
+#include "software/new_geom/polynomial.h"
+
+class Line final : public Polynomial
+{
+   public:
```
Comment (MathewMacDougall) [inline src/software/new_geom/line.h:11]: Can we also add a `slope()` function to get the slope of the line?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/line.h (hunk 3 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: This was previously discussed here:
https://github.com/UBC-Thunderbots/Software/pull/942#discussion_r336836685
https://github.com/UBC-Thunderbots/Software/pull/993#discussion_r339379819

An alternative could be to create a method in `util` to handle creating a line based on two points, and also handle the case where the two points have the same `x` value.

File: src/software/new_geom/line.h (hunk 2 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: I did it because it's a pretty short class, but I'll make a separate `.cpp` file.

File: src/software/new_geom/line.h (hunk 4 above)
Comment (jsn0716) [reply src/software/new_geom/line.h:11]: The slope can be accessed at the moment using the method inherited from `Polynomial`, `getCoeff()`. If we are adding `slope()` to this class, should we also add a `yIntercept()` function? And what about setters?

Comment (jsn0716) [reply src/software/new_geom/line.h:11]: And #1024 might be a bit relevant here, is there a convention for when to prefix a getter with `get`?

File: src/software/new_geom/line.h (hunk 1 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Addressed in 746b91a

File: src/software/new_geom/line.h (hunk 2 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Addressed in 746b91a

File: src/software/new_geom/line.h (hunk 4 above)
Comment (garethellis0) [reply src/software/new_geom/line.h:11]: I agree with Mat that we should have a `getSlope` function, and I think you make a good point about a `getYIntercept` as well (might as well have both). #1024 is in my backlog, but I think that for this class, where we won't call the methods _that_ often, we should call them `getSlope` and `getYIntercept`.

File: src/software/new_geom/line.h (hunk 1 above)
Comment (garethellis0) [reply src/software/new_geom/line.h]: Hmm, does this violate the [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)? I believe it's a violation for a subclass to accept a smaller set of arguments then it's super class, as this would mean that we could not substitute the subclass in-place of the super class in all cases. I also don't think that it's reasonable for a polynomial to be a subclass of a line (what does calling `getSlope()` mean for a polynomial?). So composition might be the way to go with this.

We should definitely chat about this today though!

Comment (garethellis0) [reply src/software/new_geom/line.h]: More understandable link: https://stackify.com/solid-design-liskov-substitution-principle/

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Main concern here is violating liskov substitution principle, see my comment reply above.

File: src/software/new_geom/line.h (hunk 1 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Yeah it does, didn't realize we were trying to stick to the principle. In that case, unless we make the classes immutable, `Line` can't be a true subtype of `Polynomial`.

Comment (scveloso) [reply src/software/new_geom/line.h]: Hmm maybe we can have ```Line``` wrap a ```Polynomial``` with an order of 1 instead

Comment (jsn0716) [reply src/software/new_geom/line.h]: As in composition? Also private inheritance could work as well

Comment (jsn0716) [reply src/software/new_geom/line.h]: Addresed in 31c130a36c3b283ceeb3119c01c817db7f16bde2. I ended up choosing composition over private inheritance since StackOverflow suggests it should be preferred. I also thought that at this point `Line` being a standalone versus composition doesn't make much of a difference, but still maintaining `Polynomial` as the underlying `Line` representation seems to make more sense.

File: src/software/new_geom/line.h (hunk 4 above)
Comment (jsn0716) [reply src/software/new_geom/line.h:11]: Addressed in 31c130a36c3b283ceeb3119c01c817db7f16bde2, `Line` class no longer has `getCoeff` and `setCoeff` methods, but instead getters and setters for slope and y-intercept.

**Review by scveloso — DISMISSED** [review]
LGTM

File: src/software/new_geom/polynomial.h [hunk 5]
```diff
@@ -45,7 +45,7 @@ class Polynomial
      * @param order the order of the term to set the coefficient
      * @param coeff the coefficient
      */
-    void setCoeff(unsigned int order, double coeff);
+    virtual void setCoeff(unsigned int order, double coeff);
```
Comment (jonathanlew) [inline src/software/new_geom/polynomial.h]: I think you can revert this change since line doesn't inherit polynomial

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good, just a minor comment

File: src/software/new_geom/line.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#include "software/new_geom/line.h"
+
+Line::Line() {}
+
+Line::Line(double y_intercept, double slope) : line(Polynomial({y_intercept, slope})) {}
```
Comment (MathewMacDougall) [inline src/software/new_geom/line.cpp]: It seems like with the interface of `y_intercept, slope` we can't create lines with infinite slope / running along the y-axis? I think this is something we would reasonably want, but it doesn't seem like polynomial can handle that anyway. Maybe we can't even use Polynomial composition here?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/polynomial.h (hunk 5 above)
Comment (jsn0716) [reply src/software/new_geom/polynomial.h]: Fixed in 23c9b47

File: src/software/new_geom/line.cpp (hunk 6 above)
Comment (jsn0716) [reply src/software/new_geom/line.cpp]: Seems like the old `Line` class doesn't handle this in a special way either, where if you call `slope` for a vertical `Line` it would just return `inf`. I believe then this would work similarly with this new `Line` class? But conceptually it doesn't make sense for a polynomial line to have an infinite slope, and evaluating it with `valueAt` wouldn't make much sense either. If handling infinite slope is important, we could just use the old representation of `Line` with two `Point`s.

Comment (garethellis0) [reply src/software/new_geom/line.cpp]: I'm inclined to say that if we want a 2D line like this, we should use two lines parameterized as a function of some third variable. ie. `x(t)` and `y(t)`.

File: src/software/new_geom/line.h (hunk 1 above)
Comment (garethellis0) [reply src/software/new_geom/line.h]: AH yeah, in general it's good to follow the [SOLID](https://scotch.io/bar-talk/s-o-l-i-d-the-first-five-principles-of-object-oriented-design) principles for OOP stuff.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - pretty simple and clean

**Review by jonathanlew — DISMISSED** [review]
LGTM!

File: src/software/new_geom/line.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/software/new_geom/line.cpp]: I think it is important to be able to represent infinite lines at _all_ orientations. It would be totally reasonable to want a line running along the y axis (maybe we're checking something with the goal-line or halfway line).

Most of the useful things we do with lines are with utility functions (that I don't think rely too much on the slope), so maybe we don't actually need access to the slope? I checked and the only place it is used is in unit tests for the ball filter, which can probably be changed to use something else.

Ideas / thoughts on making the lines able to be vertical? Maybe use a more mathematical ilne representation? Sorry for dragging this out just want to make sure we end up with the implementation that suits our needs the best.

Comment (garethellis0) [reply src/software/new_geom/line.cpp]: All good. If we want lines orientable in all directions, `y = mx + b` is not an appropriate representation. We should instead consider a line of the form indicated above, ie `f(t) = <x(t),y(t)>`. But this is theoretically just a composition of two instances of the `Line` class as it stands now. Perhaps this is a naming issue? We could call this class `SlopeInterceptLine` (or something else), and then a `Line` is the `f(t)` indicated above?

Comment (MathewMacDougall) [reply src/software/new_geom/line.cpp]: I'm not even sure "Slope intercept" line would be useful because of how we generally use this geometry on the field, the y-intercept isn't that useful, and slope limits our orientations (in a way). Sorry I didn't think of this earlier.

Comment (garethellis0) [reply src/software/new_geom/line.cpp]: I'm fine with just having a `Line` class that is defined by the `f(t)` above. Sorry for all the hassle with this @jsn0716 - does this make sense?

Comment (jsn0716) [reply src/software/new_geom/line.cpp]: Yep makes sense. So after this change, if a user wants a line with slope + y-intercept then they'll just create a first order `Polynomial`.

Comment (garethellis0) [reply src/software/new_geom/line.cpp]: Yes, I think that's reasonable.

Comment (jsn0716) [reply src/software/new_geom/line.cpp]: Finally got around to updating this, sorry for the delay! The implementation has been changed to a parametric representation. Since we are wanting `Line` to be able to represent a vertical line, it might make sense to change `getSlope` to `getDirection` or `getAngle` instead. Let me know what you think

**Comment by jsn0716** [comment]
The implementation has changed as a result of discussion here: https://github.com/UBC-Thunderbots/Software/pull/1038#discussion_r352023539. Please re-review!

File: src/software/new_geom/line.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+{
+    return second;
+}
+
+double Line::getSlope() const
+{
+    return (second.y() - first.y()) / (second.x() - first.x());
+}
+
+void Line::setFirst(const Point &first)
```
Comment (garethellis0) [inline src/software/new_geom/line.cpp]: would suggest changing to `setFirstPoint` just to be super clear, same goes for `setSecond` below.

File: src/software/new_geom/line.h [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Sets the second Point
+     *
+     * @param second the new second Point
+     */
+    void setSecond(const Point& second);
+
+    /**
+     * Calculates the Point evaluated at val, starting from the first Point and moving val
+     * units along the Line towards the second Point
```
Comment (garethellis0) [inline src/software/new_geom/line.h]: could we change this so 0 is at the first point and 1 is at the second point? Just makes some operations (like evaluating N equally spaced
 points along a line between two points) easier

File: src/software/new_geom/polynomial.cpp [hunk 9]
```diff
@@ -33,7 +33,7 @@ unsigned int Polynomial::getOrder() const
 {
     if (coeffs.size() != 0)
     {
-        for (size_t i = coeffs.size() - 1; i >= 0; i++)
+        for (size_t i = coeffs.size() - 1; i >= 0; i--)
```
Comment (garethellis0) [inline src/software/new_geom/polynomial.cpp:36]: Could we add a unit test that fails without this change to make sure we don't regress in the future?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, seems like pretty much what I expected.

File: src/software/new_geom/line.h (hunk 8 above)
Comment (jonathanlew) [reply src/software/new_geom/line.h]: This is a line so it has infinite length so does it even sense to have a first and second point? I'd rather see point and vector

File: src/software/new_geom/line.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+   private:
+    /**
+     * The first point that defines the starting point of the Line
+     */
+    Point first;
+
+    /**
+     * The second point that defines the direction of the Line
+     */
+    Point second;
```
Comment (jonathanlew) [inline src/software/new_geom/line.h]: What if this was a vector instead? Sort like ray but bidirectional

File: src/software/new_geom/line.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+     * @return the first Point
+     */
+    Point getFirst() const;
+
+    /**
+     * Returns the second Point
+     *
+     * @return the second Point
+     */
+    Point getSecond() const;
```
Comment (jonathanlew) [inline src/software/new_geom/line.h]: Line is infinitely long so getting the first and second point doesn't make a lot of sense

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Didn't realize we were going in this direction with this... Line represents an infinitely long straight line, so having points defined on it or parameterizing it with time doesn't make sense to me. What we need to know from Line is intersections, essentially and I think time parameterized polynomials actually makes that harder (search polynomial for a t that matches a given xy??). I'd like to see something ray with a point and a direction, but we can normalize the point to cross either the x and y axis (for vertical lines) and we declare that the direction goes both ways.

**Comment by jsn0716** [comment]
> Didn't realize we were going in this direction with this... Line represents an infinitely long straight line, so having points defined on it or parameterizing it with time doesn't make sense to me. What we need to know from Line is intersections, essentially and I think time parameterized polynomials actually makes that harder (search polynomial for a t that matches a given xy??). I'd like to see something ray with a point and a direction, but we can normalize the point to cross either the x and y axis (for vertical lines) and we declare that the direction goes both ways.

The parameter here doesn't necessarily have to be time, it's just a paramter `t`. I also don't think intersection would be any more difficult just because it's in a parametric form, it's just a line after all. What do you mean by normalizing the point?

File: src/software/new_geom/line.h (hunk 10 above)
Comment (garethellis0) [reply src/software/new_geom/line.h]: So I don't think it matters how we represent this internally, but I agree with your points on the interface (see below comments).

File: src/software/new_geom/line.h (hunk 11 above)
Comment (garethellis0) [reply src/software/new_geom/line.h]: Agreed, we should just be able to query a point on the line. But this gets tricky because it gets very difficult to get some context for where `t=0` is in 2D space.

Comment (jsn0716) [reply src/software/new_geom/line.h]: New implementation removes these accessors

File: src/software/new_geom/line.h (hunk 10 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Changed internal representation

File: src/software/new_geom/line.h (hunk 8 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Changed internal representation, so discussion is no longer relevant

File: src/software/new_geom/line.cpp (hunk 7 above)
Comment (jsn0716) [reply src/software/new_geom/line.cpp]: No longer relevant

**Comment by jsn0716** [comment]
Another iteration of the `Line` class, where it is now represented in a general form `ax + by + c = 0`. The main usage of `Line` in code is doing intersection / dist operations, so accessors and setters aren't necessary. There was one usage of comparing lines in doing linear regression, and `toNormalUnitVector` can be used for that.

File: src/software/new_geom/polynomial.cpp (hunk 9 above)
Comment (jsn0716) [reply src/software/new_geom/polynomial.cpp:36]: Added `get_order` test

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - thanks for persevering with this one, definitely dragged out a bit

**Review by jonathanlew — DISMISSED** [review]
Thanks for the quick turnaround!

File: src/software/new_geom/line_test.cpp [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <gtest/gtest.h>
+
+TEST(LineTest, two_points_constructor)
+{
+    Line l(Point(-1.0, 2.0), Point(3.0, 4.0));
+    EXPECT_EQ(l.toNormalUnitVector(), Vector(-1, 2).normalize());
+}
+
+TEST(LineTest, swapXY)
```
Comment (MathewMacDougall) [inline src/software/new_geom/line_test.cpp:24]: Can we add a test for getting the normal vector?

File: src/software/new_geom/line.h [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    Vector toNormalUnitVector();
+
+    /**
+     * Reflects the Line about the line y = x
+     */
+    void swapXY();
+
+   private:
+    // Coefficients for a line in the form coeffs[0]*x + coeffs[1]*y + coeffs[2] = 0
+    std::vector<double> coeffs;
```
Comment (MathewMacDougall) [inline src/software/new_geom/line.h]: nit: Because we know how many coefficients we're storing and that number won't change, we could use an std::array here

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Minor comments, thanks for the quick turnaround

File: src/software/new_geom/line_test.cpp (hunk 12 above)
Comment (jsn0716) [reply src/software/new_geom/line_test.cpp:24]: I'm not sure what that would test beyond what's being tested by `two_points_constructor` test; is there a particular type of line that you would like to see tested?

File: src/software/new_geom/line.h (hunk 13 above)
Comment (jsn0716) [reply src/software/new_geom/line.h]: Good point, fixed

File: src/software/new_geom/line_test.cpp (hunk 12 above)
Comment (MathewMacDougall) [reply src/software/new_geom/line_test.cpp:24]: Ah I didn't realize the `two_points_constructor` test used it. I think it would still be good to have a quick sanity check that the returned normal unit vector is what's expected

Comment (MathewMacDougall) [reply src/software/new_geom/line_test.cpp:24]: Doesn't have to be anything crazy complicated

Comment (jsn0716) [reply src/software/new_geom/line_test.cpp:24]: Added a test where two different `Line`s' normal unit vectors are compared. Also added a test for a vertical `Line`.

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1:

**Comment by jsn0716** [comment]
@garethellis0 @jonathanlew Added a couple tests, please re-review! As a side note, maybe automatically dismissing stale reviews with a push can be disabled? Has this been discussed previously?

**Comment by jsn0716** [comment]
@MathewMacDougall Sorry forgot to rerun clang format, please re-approve!

**Review by MathewMacDougall — APPROVED** [review]

**Review by jonathanlew — APPROVED** [review]
